drivers/vdpa/virtio/virtio_vdpa.c

```diff
@@ -0,0 +1,955 @@
+	if (ret)
+		return ret;
+
+	ret =virtio_pci_dev_interrupts_enable(priv->vpdev, vq.callfd, index+1);
+	if (ret) {
+		 DRV_LOG(ERR, "Virtq interrupt enable failed");
+		 return ret;
+	}
+
+	gpa = virtio_vdpa_hva_to_gpa(vid, (uint64_t)(uintptr_t)vq.desc);
+	if (gpa == 0) {
+		DRV_LOG(ERR, "Virtio vdpa dev %s fail to get GPA for descriptor ring %d.",
+						priv->vdev->device->name, index);
+		return -EINVAL;
+	}
+	DRV_LOG(DEBUG, "Virtq %d desc addr%"PRIx64, index, gpa);
+	priv->vrings[index]->desc = gpa;
+	vring_info.desc = gpa;
+
+	gpa = virtio_vdpa_hva_to_gpa(vid, (uint64_t)(uintptr_t)vq.avail);
+	if (gpa == 0) {
+		DRV_LOG(ERR, "Fail to get GPA for available ring.");
+		return -EINVAL;
+	}
+	DRV_LOG(DEBUG, "Virtq %d avail addr%"PRIx64, index, gpa);
+	priv->vrings[index]->avail = gpa;
+	vring_info.avail = gpa;
+
+	gpa = virtio_vdpa_hva_to_gpa(vid, (uint64_t)(uintptr_t)vq.used);
+	if (gpa == 0) {
+		DRV_LOG(ERR, "Fail to get GPA for used ring.");
+		return -EINVAL;
+	}
+	DRV_LOG(DEBUG, "Virtq %d used addr%"PRIx64, index, gpa);
+	priv->vrings[index]->used = gpa;
+	vring_info.used = gpa;
+
+	//TO_DO: need to check vq_size not exceed hw limit
+	priv->vrings[index]->size = vq.size;
+	vring_info.size = vq.size;
+
+	DRV_LOG(DEBUG, "Virtq %d nr_entrys:%d",index, vq.size);
+	if (virtio_pci_dev_queue_set(priv->vpdev, index, &vring_info)) {
+			DRV_LOG(ERR, "Setup_queue failed");
+			return -EINVAL;
+	}
+
+	ret = virtio_vdpa_virtq_doorbell_relay_enable(priv, index);
+	if (ret) {
+		 DRV_LOG(ERR, "Virtq doorbell relay failed");
+		 return ret;
+	}
+
+	priv->vrings[index]->enable = 1;
+	virtio_pci_dev_queue_notify(priv->vpdev, index);
+	return 0;
+}
+
+static int
+virtio_vdpa_vring_state_set(int vid, int vring, int state)
+{
+	struct rte_vdpa_device *vdev = rte_vhost_get_vdpa_device(vid);
+	struct virtio_vdpa_priv *priv =
+		virtio_vdpa_find_priv_resource_by_vdev(vdev);
+	int ret = 0;
+
+	if (priv == NULL) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s", vdev->device->name);
+		return -ENODEV;
+	}
+	if (vring >= (int)priv->hw_nr_virtqs) {
+		DRV_LOG(ERR, "Too big vring id: %d.", vring);
+		return -E2BIG;
+	}
+
+	//TO_DO: check if vid set here is sutiable
+	priv->vid = vid;
+
+	if(virtio_pci_dev_get_status(priv->vpdev)&VIRTIO_CONFIG_STATUS_DRIVER_OK) {
+		DRV_LOG(ERR, "Can not set vring stat when driver ok vDPA device: %s.", vdev->device->name);
+		return -EINVAL;
+	}
+
+	pthread_mutex_lock(&priv->vq_config_lock);
+	if (!state && priv->vrings[vring]->enable) {
+		ret = virtio_vdpa_virtq_disable(priv, vring);
+	}
+	if (state && !priv->vrings[vring]->enable) {
+		ret = virtio_vdpa_virtq_enable(priv, vring);
+	}
+	if (state && priv->vrings[vring]->enable) {
+		ret = virtio_vdpa_virtq_disable(priv, vring);
+		if(ret) {
+			DRV_LOG(ERR, "Fail to set vring state to 0, ret:%d vring:%d state:%d.", ret, vring, state);
+			pthread_mutex_unlock(&priv->vq_config_lock);
+			return ret;
+		}
+		ret = virtio_vdpa_virtq_enable(priv, vring);
```

Comment:
from this line, it looks like we never disable the vq.

Reply:
in 	if (!state && priv->vrings[vring]->enable) {
we disable vq